docs/qos_enforcement.md

```diff
@@ -0,0 +1,254 @@
+ # Aggregated Maximum bit rate
+ # for non-GBR bearers, this limits the aggregate bit rate
+ # Enforce this on pipelined
+ apn_ambr:
+   apn_ambr_dl: uint32
+   apn_ambr_ul: uint32
+```
+
+As an initial goal, we would like to support apn_ambr.
+
+```
+apn_configuration:
+  supported_qos_profiles: array[qos_profile]
+  # service_type should match the name of a specified ApnServiceType
+  service_type: string
+```
+
+Multiple QoS objects will be allowed to be associated to APN objects.
+
+One can be specified as the APN-level default policy. This is checked by verifying a non-zero value for fields *apn_ambr_dl* and *apn_ambr_ul* for the default QoS object.
+
+Additional QoS objects can be associated to the APN for additional services beyond the default policy. For fields *apn_ambr_dl* and *apn_ambr_ul*, these field values must be zero.
+
+```
+apn_service_type:
+  id: string
+  # Determines the IPs that will be allocated to connected subscribers under
+  # this APN service type
+  allocated_ip_range: string
+  # Specify either VLAN or the interface name for backhaul
+  vlan_id: string
+  interface: string
+```
+
+```
+lte_network:
+  id: string
+  name: string
+  description: string
+  ...
+  service_types: array[apn_service_type]
+  ...
+```
+
+```
+feg_lte_network:
+  id: string
+  name: string
+  description: string
+  ...
+  service_types: array[apn_service_type]
+  ...
+```
+
+For both LTE and federated LTE networks, service types must be associated.
+
+```
+subscriber:
+  id: string
+  name: string
+  ...
+  active_base_names: array[base_name]
+  active_policies: array[policy_id]
+  active_apns: array[string]
+  ...
+```
+
+One or more APN objects will be allowed to be associated to subscribers. This is similar to before, though will now create restrictions on associating policies to subscribers.
+
+As before, one or more policies will be allowed to be associated to subscribers. A new limitation will be created where each policy that is associated to a subscriber must share at least one of its QosProfile’s APNs with the subscriber. This must hold true for each policy to subscriber association, and CRUD operations which could break this rule should return an API error and fail to complete.
+
+```
+policy_rule:
+  id: string
+  flow_list: array[flow_description]
+  priority: uint32
+  rating_group: uint32
+  monitoring_key: string
+  ...
+  qos: array[qos_profile]
+```
+
+Multiple QoS objects will be allowed to be associated to each policy object. To implement this, its *qos* field will be changed from the current *flow_qos* type to the new QoS object type, as defined above.
+
+
+### Pipelined Changes for Enforcement
+
+At a high level, changes in `pipelined` are required to support the following:
+* (P1) Enforce QoS via AMBR across the bearer level, APN level, and subscriber level
+* (P2) Load/KPI reporting (to be used for admission control and session lifecycle management by sessiond)
+* (P3) GBR enforcement
+
```

Comment:
any changes to proto files?

Reply:
I think that would be for either Nick or Pravin to look into